### PR TITLE
Manual: replace PrioQueue example by a simpler example: FIFO queues

### PR DESCRIPTION
**Pull request by xavierleroy**
This is an alternative to #12197.  Instead of fixing the complexity bug in the priority queue example, we can replace it with a simpler example, namely a purely functional queue (FIFO).  This makes chapter 2 slightly easier to read.

Note: I named the module `Fifo` and not `Queue` to avoid confusion with the `Queue` standard library module.

Note 2: this PR includes a few minor edits elsewhere in chapter 2. I have some more edits in my mind.

### CODE REVIEW

File: manual/src/tutorials/moduleexamples.etex
Comment (avsm) [inline manual/src/tutorials/moduleexamples.etex]: Implementation looks correct to me, except for the dreaded warning 9 activating for the incomplete record pattern matches (on by default when cutting-and-pasting this into a dune project in development mode).

```
Error (warning 9 [missing-record-field-pattern]): the following labels are not bound in this record pattern:
rear
```

Would it be slightly clearer to just use a tuple for `'a queue` instead of a record here, since the purpose is to demonstrate the module system?

Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex]: For the record: I think it is stupid to enable warning 9 by default.  Now, since this ship has sailed a while ago, I just pushed a version that does not trigger warning 9.  I tried using a tuple, but then the inferred types don't mention `'a queue` and are full of `'a list * 'a list`, which is not nice.

File: manual/src/tutorials/moduleexamples.etex
Comment (avsm) [inline manual/src/tutorials/moduleexamples.etex]:
```suggestion
can extend the "FIFO" signature with the "top_opt" and "pop_opt"
```

**Review by avsm — APPROVED** [review]
Implementation and revised text looks correct to me (sans a minor LaTeX-typo present in previous version).

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex:93]: This can be more simply written using ordinary label qualification `{ Fifo.front = ...; rear = ... }`, but I guess it is OK for the purposes of illustrating the local open in patterns.

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex:107]: Why not write these with `try ... with`?

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex]: Should this be simply `make`?

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex:119]: This was called `make_queue` before, but perhaps it should be called `make`, in which case there is nothing to change here.

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex:120]: This hasn't been changed from the old text, but I would change `representation` to `definition` which is a bit less vague.
```suggestion
abstract (by not providing its actual definition as a concrete type).
```

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex:121]:
```suggestion
This ensures that users of the "FIFO" signature cannot violate data
```

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex]:
```suggestion
is empty, the rear list is also empty''.
```

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex]: "restrict by" or "restrict to"? The second sounds better to me. 
```suggestion
Restricting the "Fifo" structure to this signature results in
```

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex:155]: Not touched by this PR, but for clarity, I would rewrite as follows (the reference to the "current" siganture is not very clear):
```suggestion
As for modules, it is possible to copy the components of a signature inside another signature by using an "include" statement. For instance, we
```

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [inline manual/src/tutorials/moduleexamples.etex]:
```suggestion
identifiers defined inside the module into the scope of the current
```

**Review by nojb — APPROVED** [review]
LGTM modulo a few tiny issues.

**Comment by nojb** [comment]
I forgot to mention in the review: I prefer this option to #12197, as it is simpler without being trivial and it nicely illustrates the relevant concepts.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex:93]: Yes, the original example (with the prio queue) was also intended to illustrate local open in patterns.  Also, I prefer `Fifo.{ ... }` (a local open for the whole record) to the trick of qualifying one record field and not the others.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex:107]: I think it's easier to understand this way, esp. for readers coming from a Java/C++ background where exceptions are expensive.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex]: Right, I need to make up my mind :-)  OK for using `make` throughout the chapter.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex:119]: OK, let's go for `make`.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex:120]: Type abstraction used to be called "representation hiding", in that it hides the actual in-memory representation of values of the type.  I kind of like this intuition.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex:121]: In this sentence we're talking about the data structure invariants of the specific `Fifo` module, not about any invariant that any implementation of the `FIFO` signature could have.  I'd rather stick with the original wording.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex]: I would agree with "is empty as well", or just "is empty", but "is also empty" feels harder to read.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex]: OK for "restrict to".

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex:155]: Agreed with the clarification, but "as for modules" means "as regards modules" or "as far as modules are concerned", so it should be "as with modules".

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex]: Agreed.

**Comment by xavierleroy** [comment]
> I prefer this option to https://github.com/ocaml/ocaml/pull/12197,

So do I.  But we need to take a decision quickly because I'm not going to edit two versions all the way to the end, just to throw one away.

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [reply manual/src/tutorials/moduleexamples.etex:107]: Makes sense, thanks.

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [reply manual/src/tutorials/moduleexamples.etex:120]: OK, makes sense.

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [reply manual/src/tutorials/moduleexamples.etex:155]: I agree.

File: manual/src/tutorials/moduleexamples.etex
Comment (nojb) [reply manual/src/tutorials/moduleexamples.etex]: "is empty as well" sounds OK to me. I asked a native English speaker who suggested "must also be empty" or "must also be so" as possibilities. Or we could just stick to the original (I had just found it a bit casual).

**Comment by avsm** [comment]
I too prefer this version to #12197 for the manual, because the "trick" of using two lists with an invariant is a lovely illustration of how to build purely functional datastructures to a beginner audience. It's what we teach in our first-year Foundations of CS course, and it goes down very well as the first non-trivial data-structure after basic lists and tuples and before ADTs are introduced.

File: manual/src/tutorials/moduleexamples.etex
Comment (Octachron) [inline manual/src/tutorials/moduleexamples.etex:16]: Nitpicking: I am partial to the style rule to always explain abbreviations on first use.

**Comment by Octachron** [comment]
It seems to me that the consensus is in favor of this PR compared to https://github.com/ocaml/ocaml/pull/12197 .
I will cherry-pick the change to 5.1 as a documentation update once merged.

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex]: OK for "must also be empty".  Thanks!

File: manual/src/tutorials/moduleexamples.etex
Comment (xavierleroy) [reply manual/src/tutorials/moduleexamples.etex:16]: I tried but it feels clumsy.  French newspapers don't explain what SNCF means.  Likewise, I expect readers of this chapter to know what FIFO means.

**Comment by xavierleroy** [comment]
Merged and ready for cherry-picking.

**Pull request merged**
